fraud/bad_encoding.go

```diff
@@ -0,0 +1,219 @@
+				NamespaceID should be replaced with ParitySharesNamespaceID for
+				all shares except Q0(all erasure coded shares).
+				For EDS 4x4 Q0 is [[0;0],[0;1],[1:0],[1;1]].
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
+			if index >= len(errShares)/2 || int(position) >= len(errShares)/2 {
+				namespaceID = consts.ParitySharesNamespaceID
+			}
+
+			shares[index] = &Share{namespaceID, share, &proof}
+		}
+	}
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares,
+		isRow:       isRow,
+		Position:    position,
+	}, nil
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to []byte
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEnconding{
+		Height:   p.BlockHeight,
+		Shares:   shares,
+		Position: uint32(p.Position),
+		IsRow:    p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
+	in := pb.BadEnconding{}
+	if err := in.Unmarshal(data); err != nil {
+		return err
+	}
+	befp := UnmarshalBefp(&in)
+	*p = *befp
+
+	return nil
+}
+
+func UnmarshalBefp(befp *pb.BadEnconding) *BadEncodingProof {
+	return &BadEncodingProof{
+		BlockHeight: befp.Height,
+		Shares:      ProtoToShare(befp.Shares),
+	}
+}
+
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader, codec rsmt2d.Codec) error {
+	merkleRowRoots := header.DAH.RowsRoots
+	merkleColRoots := header.DAH.ColumnRoots
+	if int(p.Position) >= len(merkleRowRoots) || int(p.Position) >= len(merkleColRoots) {
+		return errors.New("invalid fraud proof: incorrect position of bad row/col")
+	}
+	if len(merkleRowRoots) != len(merkleColRoots) {
+		return errors.New("invalid fraud proof: invalid extended header")
+	}
+	if len(merkleRowRoots) != len(p.Shares) || len(merkleColRoots) != len(p.Shares) {
+		return errors.New("invalid fraud proof: invalid shares")
+	}
+
+	roots := merkleRowRoots
+	if p.isRow {
+		roots = merkleColRoots
+	}
+
+	shares := make([][]byte, len(roots))
+
+	// verify that Merkle proofs correspond to particular shares
+	for index, share := range p.Shares {
+		if share == nil {
+			continue
+		}
+		shares[index] = share.Raw
+		if ok := share.Validate(roots[index]); !ok {
+			return fmt.Errorf("invalid fraud proof: incorrect share received at position %d", index)
+		}
+	}
+
+	// rebuild a row/col
+	rebuiltShares, err := codec.Decode(shares)
+	if err != nil {
+		return err
+	}
+	rebuiltExtendedShares, err := codec.Encode(rebuiltShares)
+	if err != nil {
+		return err
+	}
+	rebuiltShares = append(rebuiltShares, rebuiltExtendedShares...)
```

Comment:
Double-check that this works the same with Leopard, IIRC Leopard behaves differently with that it returns, but this might be handled by the wrapper.

Reply:
IIRC, I've updated leopard/the go-wrapper to behave the same with this regard but would be good to confirm this behaviour.

Reply:
We can do this in followup work. I think we should simply switch to leopard by default in rsmt2d and core. Then, this code will automatically test with leopard once we updated these deps.

Reply:
see: https://github.com/celestiaorg/rsmt2d/issues/74 and https://github.com/celestiaorg/rsmt2d/pull/75